File: solutions/Golang/word_break_dp_trie.go
package main

import "fmt"

type TrieNode struct {
	children map[rune]*TrieNode
	is_end   bool
}

func NewTrieNode() *TrieNode {
	return &TrieNode{
		children: make(map[rune]*TrieNode),
		is_end:   false,
	}
}

/*
 * Determine if string can be segmented using words from dictionary (Trie approach).
 *
 * Time Complexity: O(n*m)
 * Space Complexity: O(n + T)
 */
func wordBreakDpTrie(s string, wordDict []string) bool {
	root := NewTrieNode()

	for _, word := range wordDict {
		node := root
		for _, ch := range word {
			if _, exists := node.children[ch]; !exists {
				node.children[ch] = NewTrieNode()
			}
			node = node.children[ch]
		}
		node.is_end = true
	}

	chars := []rune(s)
	dp := make([]bool, len(chars)+1)
	dp[0] = true

	for i := 1; i <= len(chars); i++ {
		if !dp[i] {
			node := root
			for j := i - 1; j >= 0; j-- {
				ch := chars[j]
				if next, exists := node.children[ch]; exists {
					node = next
					if dp[j] && node.is_end {
						dp[i] = true
						break
					}
				} else {
					break
				}
			}
		}
	}

	return dp[len(chars)]
}

func main() {
	fmt.Println(wordBreakDpTrie("leetcode", []string{"leet", "code"}))  // true
	fmt.Println(wordBreakDpTrie("applepenapple", []string{"apple", "pen"}))  // true
}

File: solutions/Python/word_break_dp_trie.py
from typing import List


class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end = False


def word_break_dp_trie(s: str, word_dict: List[str]) -> bool:
    """
    Determine if string can be segmented using words from dictionary (Trie approach).

    Time Complexity: O(n*m) where n = len(s), m = max word length
    Space Complexity: O(n) for DP array + O(T) for Trie where T = total characters in dictionary

    Build a Trie from the dictionary, then use DP with Trie for faster lookups.
    dp[i] = True if s[0:i] can be segmented.
    Instead of checking all substrings (O(n²)), we traverse the Trie (O(m)).
    """
    # Build Trie
    root = TrieNode()
    for word in word_dict:
        node = root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end = True

    dp = [False] * (len(s) + 1)
    dp[0] = True

    for i in range(1, len(s) + 1):
        if not dp[i]:
            # Try to find a word ending at position i
            node = root
            for j in range(i - 1, -1, -1):
                char = s[j]
                if char not in node.children:
                    break
                node = node.children[char]
                if dp[j] and node.is_end:
                    dp[i] = True
                    break

    return dp[len(s)]


if __name__ == "__main__":
    print(word_break_dp_trie("leetcode", ["leet", "code"]))           # True
    print(word_break_dp_trie("applepenapple", ["apple", "pen"]))      # True
    print(word_break_dp_trie("catsandog", ["cat", "cats", "and", "sand", "dog"]))  # False
